src/nvim/api/buffer.c

```diff
@@ -1465,15 +1465,21 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
   }
 
   if (col2 >= 0) {
-    if (line2 >= 0) {
-      len = STRLEN(ml_get_buf(buf, (linenr_T)line2+1, false));
+    if (line2 >= 0 && line2 != buf->b_ml.ml_line_count) {
+      len = STRLEN(ml_get_buf(buf, (linenr_T)line2 + 1, false));
+    } else if (line2 == buf->b_ml.ml_line_count) {
+      // We are trying to add an extmark past final newline
+      if (col2 != 0) {
+        // Error out
+        col2 = MAXCOL;
+      }
     } else {
       // reuse len from before
       line2 = (int)line;
     }
-    if (col2 > (Integer)len) {
-      api_set_error(err, kErrorTypeValidation,
-                    "end_col value outside range");
+    // We can set extmarks past newlines
+    if (col2 > (Integer)len + 1) {
```

Comment:
I don't think this is correct. The point was that "past newline" state is marked as col=0 at the _next_ line, which is why the line_count check needed update.